### DISCUSSION TITLE
Filter on custom attributes when calling Get-PnPAzureADUser

### DISCUSSION THREAD

**Discussion by EA12** [pnp/powershell#1494] [Q&A]
Hi experts,

how to use -Filter on a custom attribute against null values?
I am getting 'Invalid filter clause'

Example: `$users = Get-PnPAzureADUser -Select "DisplayName","extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode" -Filter "extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode -ne ''"`
or
`$users = Get-PnPAzureADUser -Select "DisplayName","extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode" -Filter "extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode -ne '$null'"`
or
`$users = Get-PnPAzureADUser -Select "DisplayName","extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode" -Filter "extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode -ne $null"`

This is working, but I don't want to fetch all user first:
`$users = Get-PnPAzureADUser -Select "DisplayName","extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode" | Where-Object { $_.AdditionalProperties["extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode"] -ne $null } 
    `
Thx, Ronny

**Comment by LeonArmston** [comment]
MS graph user properties notoriously dont work with filtering on null values.

If I was you I'd test using the MS Graph Explorer ([Added a sample linked here](https://developer.microsoft.com/en-us/graph/graph-explorer?request=users%3F%24select%3DBusinessPhones%2CDisplayName%2CGivenName%2CJobTitle%2CMail%2CMobilePhone%2COfficeLocation%2CPreferredLanguage%2CSurname%2CUserPrincipalName%2CId%2CAccountEnabled%26%24filter%3DAccountEnabled%20eq%20true&method=GET&version=v1.0&GraphUrl=https://graph.microsoft.com)) as this is what the PnP cmdlet is using behind the scenes and filtering uses the same syntax  and then see if you custom property can be filtered for null. Believe the syntax could be:

extension_b8fc35d8e8ec45e689d332303177957a_ZZZ_townCode ne null

If not I think the next best way to do it would be retrieve all users like you are trying to avoid or have a regular scheduled process that writes the user data to another data source that can be filtered for null.

**Reply by EA12** [reply]
Hi Leon,
your suggested filter syntax does not work.
    Message: **Unsupported property filter clause operator 'NotEqualsMatch'.**
It seems that this is not possible.
From [this](https://docs.microsoft.com/en-us/previous-versions/azure/ad/graph/howto/azure-ad-graph-api-supported-queries-filters-and-paging-options) doc:
null values are not supported as an operand in filter expressions. For example, you cannot specify a null value to filter for unset properties.

Thanks, anyway.

**Reply by LeonArmston** [reply]
Hi @EA12 

Yes I did say nulls are notoriously bad in MS Graph - there are some "hacky" workarounds out there but it depends on the type of field.

Think the only way to do what you want is to retrieve all users like you are trying to avoid or have a regular scheduled process that writes all the user data to another data source that can be filtered for null if you need speed.

Thanks

Leon
